
diff -up gdm-2.29.5/daemon/gdm-simple-slave.c.fix-boot gdm-2.29.5/daemon/gdm-simple-slave.c
--- gdm-2.29.5/daemon/gdm-simple-slave.c.fix-boot	2010-01-26 17:34:33.980834884 -0500
+++ gdm-2.29.5/daemon/gdm-simple-slave.c	2010-01-26 17:34:34.027782630 -0500
@@ -91,6 +91,7 @@ struct GdmSimpleSlavePrivate
 
 enum {
         PROP_0,
+        FORCE_ACTIVE_VT
 };
 
 static void     gdm_simple_slave_class_init     (GdmSimpleSlaveClass *klass);
@@ -1327,11 +1328,13 @@ gdm_simple_slave_run (GdmSimpleSlave *sl
         char    *display_name;
         char    *auth_file;
         gboolean display_is_local;
+        gboolean force_active_vt;
 
         g_object_get (slave,
                       "display-is-local", &display_is_local,
                       "display-name", &display_name,
                       "display-x11-authority-file", &auth_file,
+                      "force-active-vt", &force_active_vt,
                       NULL);
 
         /* if this is local display start a server if one doesn't
@@ -1369,7 +1372,11 @@ gdm_simple_slave_run (GdmSimpleSlave *sl
                         plymouth_prepare_for_transition (slave);
                         res = gdm_server_start_on_active_vt (slave->priv->server);
                 } else {
-                        res = gdm_server_start (slave->priv->server);
+                        if (force_active_vt) {
+                                res = gdm_server_start_on_active_vt (slave->priv->server);
+                        } else {
+                                res = gdm_server_start (slave->priv->server);
+                        }
                 }
                 if (! res) {
                         g_warning (_("Could not start the X "
@@ -1521,12 +1528,14 @@ gdm_simple_slave_finalize (GObject *obje
 }
 
 GdmSlave *
-gdm_simple_slave_new (const char *id)
+gdm_simple_slave_new (const char *id,
+                      gboolean    force_active_vt)
 {
         GObject *object;
 
         object = g_object_new (GDM_TYPE_SIMPLE_SLAVE,
                                "display-id", id,
+                               "force-active-vt", force_active_vt,
                                NULL);
 
         return GDM_SLAVE (object);
diff -up gdm-2.29.5/daemon/gdm-simple-slave.h.fix-boot gdm-2.29.5/daemon/gdm-simple-slave.h
--- gdm-2.29.5/daemon/gdm-simple-slave.h.fix-boot	2009-12-08 10:20:18.000000000 -0500
+++ gdm-2.29.5/daemon/gdm-simple-slave.h	2010-01-26 17:34:34.027782630 -0500
@@ -48,7 +48,8 @@ typedef struct
 } GdmSimpleSlaveClass;
 
 GType               gdm_simple_slave_get_type   (void);
-GdmSlave *          gdm_simple_slave_new        (const char       *id);
+GdmSlave *          gdm_simple_slave_new        (const char       *id,
+                                                 gboolean          force_active_vt);
 
 G_END_DECLS
 
diff -up gdm-2.29.5/daemon/gdm-slave.c.fix-boot gdm-2.29.5/daemon/gdm-slave.c
--- gdm-2.29.5/daemon/gdm-slave.c.fix-boot	2010-01-26 17:34:33.777750861 -0500
+++ gdm-2.29.5/daemon/gdm-slave.c	2010-01-26 17:46:49.256752923 -0500
@@ -84,6 +84,7 @@ struct GdmSlavePrivate
         char            *display_hostname;
         gboolean         display_is_local;
         gboolean         display_is_parented;
+        gboolean         force_active_vt;
         char            *display_seat_id;
         char            *display_x11_authority_file;
         char            *parent_display_name;
@@ -102,6 +103,7 @@ enum {
         PROP_DISPLAY_NUMBER,
         PROP_DISPLAY_HOSTNAME,
         PROP_DISPLAY_IS_LOCAL,
+        PROP_FORCE_ACTIVE_VT,
         PROP_DISPLAY_SEAT_ID,
         PROP_DISPLAY_X11_AUTHORITY_FILE
 };
@@ -1402,6 +1404,13 @@ _gdm_slave_set_display_is_local (GdmSlav
 }
 
 static void
+_gdm_slave_set_force_active_vt (GdmSlave   *slave,
+                                gboolean    force_active_vt)
+{
+        slave->priv->force_active_vt = force_active_vt;
+}
+
+static void
 gdm_slave_set_property (GObject      *object,
                         guint         prop_id,
                         const GValue *value,
@@ -1433,6 +1442,9 @@ gdm_slave_set_property (GObject      *ob
         case PROP_DISPLAY_IS_LOCAL:
                 _gdm_slave_set_display_is_local (self, g_value_get_boolean (value));
                 break;
+        case PROP_FORCE_ACTIVE_VT:
+                _gdm_slave_set_force_active_vt (self, g_value_get_boolean (value));
+                break;
         default:
                 G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
                 break;
@@ -1471,6 +1483,9 @@ gdm_slave_get_property (GObject    *obje
         case PROP_DISPLAY_IS_LOCAL:
                 g_value_set_boolean (value, self->priv->display_is_local);
                 break;
+        case PROP_FORCE_ACTIVE_VT:
+                g_value_set_boolean (value, self->priv->force_active_vt);
+                break;
         default:
                 G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
                 break;
@@ -1595,6 +1610,14 @@ gdm_slave_class_init (GdmSlaveClass *kla
                                                                "display is local",
                                                                TRUE,
                                                                G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+        g_object_class_install_property (object_class,
+                                         PROP_FORCE_ACTIVE_VT,
+                                         g_param_spec_boolean ("force-active-vt",
+                                                               "Force Active VT",
+                                                               "Force display to active VT",
+                                                               TRUE,
+                                                               G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
 
         signals [STOPPED] =
                 g_signal_new ("stopped",
diff -up gdm-2.29.5/daemon/gdm-static-display.c.fix-boot gdm-2.29.5/daemon/gdm-static-display.c
--- gdm-2.29.5/daemon/gdm-static-display.c.fix-boot	2009-12-17 09:20:42.000000000 -0500
+++ gdm-2.29.5/daemon/gdm-static-display.c	2010-01-26 17:34:34.029750726 -0500
@@ -86,10 +86,27 @@ gdm_static_display_remove_user_authoriza
 }
 
 static gboolean
+triggered_to_force_display_on_active_vt (void)
+{
+        gboolean should_force_display_on_active_vt;
+
+        should_force_display_on_active_vt = g_file_test (GDM_SPOOL_DIR "/force-display-on-active-vt",
+                                                         G_FILE_TEST_EXISTS);
+        g_unlink (GDM_SPOOL_DIR "/force-display-on-active-vt");
+
+        return should_force_display_on_active_vt;
+}
+
+static gboolean
 gdm_static_display_manage (GdmDisplay *display)
 {
         g_return_val_if_fail (GDM_IS_DISPLAY (display), FALSE);
 
+        if (triggered_to_force_display_on_active_vt ()) {
+                g_object_set (display, "force-active-vt", TRUE, NULL);
+        } else {
+                g_object_set (display, "force-active-vt", FALSE, NULL);
+        }
         GDM_DISPLAY_CLASS (gdm_static_display_parent_class)->manage (display);
 
         return TRUE;
diff -up gdm-2.29.5/daemon/Makefile.am.fix-boot gdm-2.29.5/daemon/Makefile.am
--- gdm-2.29.5/daemon/Makefile.am.fix-boot	2010-01-12 15:30:45.000000000 -0500
+++ gdm-2.29.5/daemon/Makefile.am	2010-01-26 17:34:34.029750726 -0500
@@ -14,6 +14,7 @@ AM_CPPFLAGS = \
 	-DLOGDIR=\"$(logdir)\"				\
 	-DSBINDIR=\"$(sbindir)\"			\
 	-DGNOMELOCALEDIR=\""$(datadir)/locale"\"	\
+	-DGDM_SPOOL_DIR=\"/var/spool/gdm\"  \
 	-DGDM_XAUTH_DIR=\"$(GDM_XAUTH_DIR)\"		\
 	-DGDM_SCREENSHOT_DIR=\"$(GDM_SCREENSHOT_DIR)\"		\
 	-DGDM_CACHE_DIR=\""$(localstatedir)/cache/gdm"\"	\
diff -up gdm-2.29.5/daemon/simple-slave-main.c.fix-boot gdm-2.29.5/daemon/simple-slave-main.c
--- gdm-2.29.5/daemon/simple-slave-main.c.fix-boot	2010-01-12 11:36:37.000000000 -0500
+++ gdm-2.29.5/daemon/simple-slave-main.c	2010-01-26 17:34:34.029750726 -0500
@@ -178,9 +178,11 @@ main (int    argc,
         DBusGConnection  *connection;
         GdmSlave         *slave;
         static char      *display_id = NULL;
+        static gboolean   force_active_vt = FALSE;
         GdmSignalHandler *signal_handler;
         static GOptionEntry entries []   = {
                 { "display-id", 0, 0, G_OPTION_ARG_STRING, &display_id, N_("Display ID"), N_("ID") },
+                { "force-active-vt", 0, 0, G_OPTION_ARG_NONE, &force_active_vt, N_("Force X to start on active vt"), NULL },
                 { NULL }
         };
 
@@ -248,7 +250,7 @@ main (int    argc,
         gdm_signal_handler_add (signal_handler, SIGUSR1, signal_cb, NULL);
         gdm_signal_handler_add (signal_handler, SIGUSR2, signal_cb, NULL);
 
-        slave = gdm_simple_slave_new (display_id);
+        slave = gdm_simple_slave_new (display_id, force_active_vt);
         if (slave == NULL) {
                 goto out;
         }
diff -up gdm-2.29.5/data/Makefile.am.fix-boot gdm-2.29.5/data/Makefile.am
--- gdm-2.29.5/data/Makefile.am.fix-boot	2010-01-13 09:44:40.000000000 -0500
+++ gdm-2.29.5/data/Makefile.am	2010-01-26 17:34:34.030752200 -0500
@@ -13,6 +13,7 @@ predir = $(gdmconfdir)/PreSession
 postlogindir = $(gdmconfdir)/PostLogin
 workingdir = $(GDM_WORKING_DIR)
 xauthdir = $(GDM_XAUTH_DIR)
+spooldir = "/var/spool/gdm"
 screenshotdir = $(GDM_SCREENSHOT_DIR)
 cachedir = $(localstatedir)/cache/gdm
 
@@ -129,6 +130,7 @@ uninstall-hook:
 	$(DESTDIR)$(workingdir)/.gconf.mandatory \
 	$(DESTDIR)$(screenshotdir) \
 	$(DESTDIR)$(xauthdir)
+	$(DESTDIR)$(spooldir)
 
 install-data-hook: gdm.conf-custom Xsession Init PostSession PreSession gconf.path
 	if test '!' -d $(DESTDIR)$(gdmconfdir); then \
@@ -221,6 +223,11 @@ install-data-hook: gdm.conf-custom Xsess
 		chmod 1770 $(DESTDIR)$(workingdir); \
 		chown root:gdm $(DESTDIR)$(workingdir) || : ; \
 	fi
+	if test '!' -d $(DESTDIR)$(spooldir); then \
+		$(mkinstalldirs) $(DESTDIR)$(spooldir); \
+		chmod 775 $(DESTDIR)$(spooldir); \
+		chown root:gdm $(DESTDIR)$(spooldir) || : ; \
+	fi
 
 	if test '!' -d $(DESTDIR)$(cachedir); then \
 		$(mkinstalldirs) $(DESTDIR)$(cachedir); \
