Index: gdm/gui/simple-greeter/gdm-simple-greeter.schemas.in
===================================================================
--- gdm/gui/simple-greeter/gdm-simple-greeter.schemas.in	(revision 5874)
+++ gdm/gui/simple-greeter/gdm-simple-greeter.schemas.in	(working copy)
@@ -6,7 +6,7 @@
       <applyto>/apps/gdm/simple-greeter/logo-icon-name</applyto>
       <owner>gdm-simple-greeter</owner>
       <type>string</type>
-      <default>computer</default>
+      <default>start-here</default>
       <locale name="C">
         <short>Icon name to use for greeter logo</short>
         <long>Set to the themed icon name to use for the greeter logo.</long>
