--- gdm-2.19.1/daemon/slave.c.clean-up-xsession-errors	2007-05-13 22:08:25.000000000 -0400
+++ gdm-2.19.1/daemon/slave.c	2007-05-21 11:27:27.000000000 -0400
@@ -4523,6 +4523,11 @@ gdm_slave_session_stop (gboolean run_pos
 
 	finish_session_output (run_post_session /* do_read */);
 
+	/* If successfully exited then clear the log file
+	 */
+	if (d->xsession_errors_filename != NULL)
+		VE_IGNORE_EINTR (unlink (d->xsession_errors_filename)
+
 	if (local_login == NULL)
 		pwent = NULL;
 	else
