diff -up gdm-2.21.2/data/session-setup.entries.fix-background gdm-2.21.2/data/session-setup.entries
--- gdm-2.21.2/data/session-setup.entries.fix-background	2007-11-19 15:42:00.000000000 -0500
+++ gdm-2.21.2/data/session-setup.entries	2007-11-19 16:05:43.000000000 -0500
@@ -103,6 +104,12 @@
       </value>
     </entry>
     <entry>
+      <key>/desktop/gnome/background/picture_filename</key>
+      <value>
+        <string>/usr/share/backgrounds/images/default.png</string>
+      </value>
+    </entry>
+    <entry>
       <key>/desktop/gnome/url-handlers/http/command</key>
       <value>
         <string>/bin/true</string>
