diff -up gdm-2.25.2/daemon/gdm-display-access-file.c.use-resolvable-hostname gdm-2.25.2/daemon/gdm-display-access-file.c
--- gdm-2.25.2/daemon/gdm-display-access-file.c.use-resolvable-hostname	2009-03-10 23:28:29.355897686 -0400
+++ gdm-2.25.2/daemon/gdm-display-access-file.c	2009-03-10 23:28:33.925897051 -0400
@@ -410,7 +410,7 @@ _get_auth_info_for_display (GdmDisplayAc
 
         if (is_local) {
                 *family = FamilyLocal;
-                *address = g_strdup (g_get_host_name ());
+                *address = g_strdup ("localhost");
         } else {
                 *family = FamilyWild;
                 gdm_display_get_remote_hostname (display, address, NULL);
