Index: data/greeter-autostart/Makefile.am
===================================================================
--- data/greeter-autostart/Makefile.am	(revision 6433)
+++ data/greeter-autostart/Makefile.am	(working copy)
@@ -19,6 +19,7 @@
 	gnome-power-manager.desktop.in		\
 	gnome-settings-daemon.desktop.in	\
 	gok.desktop.in				\
+	metacity.desktop.in			\
 	orca-screen-reader.desktop.in		\
 	$(NULL)
 desktop_DATA = $(desktop_in_files:.desktop.in=.desktop)
Index: data/greeter-autostart/metacity.desktop.in
===================================================================
--- data/greeter-autostart/metacity.desktop.in	(revision 0)
+++ data/greeter-autostart/metacity.desktop.in	(revision 0)
@@ -0,0 +1,20 @@
+[Desktop Entry]
+Type=Application
+Encoding=UTF-8
+_Name=Metacity
+Exec=metacity
+# name of loadable control center module
+X-GNOME-WMSettingsModule=metacity
+# name we put on the WM spec check window
+X-GNOME-WMName=Metacity
+# back compat only 
+X-GnomeWMSettingsLibrary=metacity
+X-GNOME-Bugzilla-Bugzilla=GNOME
+X-GNOME-Bugzilla-Product=metacity
+X-GNOME-Bugzilla-Component=general
+X-GNOME-Autostart-Phase=WindowManager
+X-GNOME-Provides=windowmanager
+X-GNOME-Autostart-Notify=true
+
+[Window Manager]
+SessionManaged=true
Index: po/POTFILES.in
===================================================================
--- po/POTFILES.in	(revision 6433)
+++ po/POTFILES.in	(working copy)
@@ -56,6 +56,7 @@
 data/greeter-autostart/gnome-settings-daemon.desktop.in.in
 data/greeter-autostart/gnome-mag.desktop.in
 data/greeter-autostart/gok.desktop.in
+data/greeter-autostart/metacity.desktop.in
 data/greeter-autostart/orca-screen-reader.desktop.in
 gui/simple-chooser/gdm-host-chooser-dialog.c
 gui/simple-chooser/gdm-host-chooser-widget.c
Index: po/ChangeLog
===================================================================
--- po/ChangeLog	(revision 6433)
+++ po/ChangeLog	(working copy)
@@ -1,3 +1,7 @@
+2008-08-25  William Jon McCann  <jmccann@redhat.com>
+
+	* POTFILES.in: Update list
+
 2008-08-25  Tomasz Dominikowski  <tdominikowski@aviary.pl>
 
 	* pl.po: Updated Polish translation
