diff -up gdm-2.21.2/daemon/gdm.in.dont-run-profile gdm-2.21.2/daemon/gdm.in
--- gdm-2.21.2/daemon/gdm.in.dont-run-profile	2007-11-15 10:14:06.000000000 -0500
+++ gdm-2.21.2/daemon/gdm.in	2007-11-15 10:14:43.000000000 -0500
@@ -1,11 +1,5 @@
 #!/bin/sh
 #
-# A script so that
-#    1) we read the standard system env vars
-#    2) syadmins/integrators can add their own private options etc...
-
-test -f /etc/profile && . /etc/profile
-
 # Try to set LC_MESSAGES to LANG if neither LC_MESSAGES or LC_ALL is set.
 # This ensures that gdm can display in a system's default language if only
 # LANG is set.
