--- gdm-2.17.1/gui/gdmlogin.c.desensitize-entry	2006-11-05 14:22:49.000000000 -0500
+++ gdm-2.17.1/gui/gdmlogin.c	2006-11-05 14:24:05.000000000 -0500
@@ -1776,7 +1776,7 @@
 	    curuser = NULL;
 	}
 
-	gtk_widget_set_sensitive (entry, TRUE);
+	gtk_widget_set_sensitive (entry, FALSE);
 	gtk_widget_set_sensitive (ok_button, FALSE);
 	gtk_widget_set_sensitive (start_again_button, FALSE);
 
--- gdm-2.17.1/gui/greeter/greeter_item_pam.c.desensitize-entry	2006-11-05 14:22:49.000000000 -0500
+++ gdm-2.17.1/gui/greeter/greeter_item_pam.c	2006-11-05 14:22:50.000000000 -0500
@@ -291,11 +291,20 @@
     {
       entry = GNOME_CANVAS_WIDGET (entry_info->item)->widget;
       
-      gtk_entry_set_visibility (GTK_ENTRY (entry), entry_visible);
-      gtk_widget_set_sensitive (GTK_WIDGET (entry), TRUE);
-      gtk_entry_set_max_length (GTK_ENTRY (entry), entry_len);
-      gtk_entry_set_text (GTK_ENTRY (entry), "");
-      gtk_widget_grab_focus (entry);
+      if (message != NULL && strcmp (message, "") != 0) 
+        {
+          gtk_entry_set_visibility (GTK_ENTRY (entry), entry_visible);
+          gtk_widget_set_sensitive (GTK_WIDGET (entry), TRUE);
+          gtk_entry_set_max_length (GTK_ENTRY (entry), entry_len);
+          gtk_entry_set_text (GTK_ENTRY (entry), "");
+          gtk_widget_grab_focus (entry);
+        }
+      else
+        {
+          gtk_entry_set_visibility (GTK_ENTRY (entry), entry_visible);
+          gtk_entry_set_text (GTK_ENTRY (entry), "...");
+          gtk_widget_set_sensitive (GTK_WIDGET (entry), FALSE);
+        }
     }
 
   messages_to_give = FALSE;
