diff -up gdm-2.19.5/gui/gdmlogin.c.disable-typeahead gdm-2.19.5/gui/gdmlogin.c
--- gdm-2.19.5/gui/gdmlogin.c.disable-typeahead	2007-08-16 10:49:46.000000000 -0400
+++ gdm-2.19.5/gui/gdmlogin.c	2007-08-16 10:49:59.000000000 -0400
@@ -2393,6 +2393,8 @@ gdm_login_gui_init (void)
 	    gtk_tree_view_set_rules_hint (GTK_TREE_VIEW (browser), FALSE);
 	    gtk_tree_view_set_headers_visible (GTK_TREE_VIEW (browser),
 					       FALSE);
+	    gtk_tree_view_set_enable_search (GTK_TREE_VIEW (browser),
+					     FALSE);
 	    selection = gtk_tree_view_get_selection (GTK_TREE_VIEW (browser));
 	    gtk_tree_selection_set_mode (selection, GTK_SELECTION_SINGLE);
 
diff -up gdm-2.19.5/gui/greeter/greeter_item_ulist.c.disable-typeahead gdm-2.19.5/gui/greeter/greeter_item_ulist.c
--- gdm-2.19.5/gui/greeter/greeter_item_ulist.c.disable-typeahead	2007-08-16 10:49:08.000000000 -0400
+++ gdm-2.19.5/gui/greeter/greeter_item_ulist.c	2007-08-16 10:49:18.000000000 -0400
@@ -333,6 +333,9 @@ greeter_generate_userlist (GtkWidget *tv
 
 	gtk_tree_view_set_headers_visible (GTK_TREE_VIEW (tv),
 					   FALSE);
+	gtk_tree_view_set_enable_search (GTK_TREE_VIEW (tv),
+					 FALSE);
+
 	selection = gtk_tree_view_get_selection (GTK_TREE_VIEW (tv));
 	gtk_tree_selection_set_mode (selection, GTK_SELECTION_SINGLE);
 	if (users != NULL) {
