--- gdm-2.6.0.8/daemon/verify-pam.c.verify-check	2005-06-21 17:10:14.000000000 +0200
+++ gdm-2.6.0.8/daemon/verify-pam.c	2005-06-21 18:49:50.000000000 +0200
@@ -1467,6 +1467,10 @@
 void
 gdm_verify_check (void)
 {
+/* This check creates spurious audit entries and it isn't necessary
+   to perform it when gdm is part of the distribution.
+*/
+#if 0
 	pam_handle_t *ph = NULL;
 
 	if (pam_start ("gdm", NULL, &standalone_pamc, &ph) != PAM_SUCCESS) {
@@ -1487,6 +1491,7 @@
 
 	closelog ();
 	openlog ("gdm", LOG_PID, LOG_DAEMON);
+#endif
 }
 
 /* used in pam */

