--- gdm-2.16.0/config/locale.alias.indic-langs	2006-09-19 11:47:50.000000000 -0400
+++ gdm-2.16.0/config/locale.alias	2006-09-19 11:49:30.000000000 -0400
@@ -20,6 +20,7 @@
 Arabic(Egypt)		ar_EG.UTF-8,ar_EG
 Arabic(Lebanon)		ar_LB.UTF-8,ar_LB
 Armenian		hy_AM.UTF-8,hy_AM
+Assamese		as_IN.UTF-8,as_IN
 Azerbaijani		az_AZ.UTF-8,az_AZ
 Azerbaijani(Iran)	az_IR.UTF-8,az_IR
 Basque			eu_ES.UTF-8,eu_ES
@@ -94,6 +95,7 @@
 Serbian			sr_CS.UTF-8,sr_CS,sr_YU@cyrillic,sr_YU
 Serbian(Latin)		sr_CS.UTF-8@Latn,sr_CS@Latn,sr_YU@Latn
 Serbian(Jekavian)	sr_CS.UTF-8@ije,sr_CS@ije,sr_YU@ije
+Sinhala			si_LK.UTF-8,si_LK
 Slovak			sk_SK.UTF-8,sk_SK
 Slovenian		sl_SI.UTF-8,sl_SI
 Spanish			es_ES.UTF-8,es_ES
--- gdm-2.16.0/gui/gdmlanguages.c.indic-langs	2006-09-19 11:43:03.000000000 -0400
+++ gdm-2.16.0/gui/gdmlanguages.c	2006-09-19 11:47:14.000000000 -0400
@@ -62,6 +62,8 @@
 	/*Note translate the A-M to the A-M you used in the group label */
 	{ N_("A-M|Armenian"), "hy_AM", "Հայերեն", 0 },
 	/*Note translate the A-M to the A-M you used in the group label */
+	{ N_("A-M|Assamese"), "as_IN", "অসমীয়া", 0 },
+	/*Note translate the A-M to the A-M you used in the group label */
 	{ N_("A-M|Azerbaijani"), "az_AZ", "Azərbaycanca", 0 },
 	/*Note translate the A-M to the A-M you used in the group label */
 	{ N_("A-M|Basque"), "eu_ES", "Euskara", 0 },
@@ -200,6 +202,8 @@
 	/*Note translate the N-Z to the N-Z you used in the group label */
 	{ N_("N-Z|Serbian (Jekavian)"), "sr_CS@ije", "српски (ијекавски)", 0 },
 	/*Note translate the N-Z to the N-Z you used in the group label */
+	{ N_("N-Z|Sinhala"), "si_LK", "සිංහල", 0 },
+	/*Note translate the N-Z to the N-Z you used in the group label */
 	{ N_("N-Z|Slovak"), "sk_SK", "Slovenský", 0 },
 	/*Note translate the N-Z to the N-Z you used in the group label */
 	{ N_("N-Z|Slovenian"), "sl_SI", "Slovenščina", 0 },
