--- gdm-2.6.0.5/daemon/slave.c.old	2004-10-20 05:58:05.449740960 -0400
+++ gdm-2.6.0.5/daemon/slave.c	2004-10-20 05:58:37.230909488 -0400
@@ -4293,7 +4293,12 @@
 #endif
 
     finish_session_output (run_post_session /* do_read */);
-    
+
+    /* If successfully exited then clear the log file
+     */
+    if (d->xsession_errors_filename != NULL)
+	    VE_IGNORE_EINTR (unlink (d->xsession_errors_filename));
+
     if (local_login == NULL)
 	    pwent = NULL;
     else
