diff -up gdm-2.25.2/daemon/gdm-server.c.append-logs gdm-2.25.2/daemon/gdm-server.c
--- gdm-2.25.2/daemon/gdm-server.c.append-logs	2009-01-22 10:42:10.043245824 -0500
+++ gdm-2.25.2/daemon/gdm-server.c	2009-01-22 10:42:30.770530113 -0500
@@ -421,7 +421,7 @@ server_child_setup (GdmServer *server)
         g_debug ("GdmServer: Opening logfile for server %s", log_path);
 
         VE_IGNORE_EINTR (g_unlink (log_path));
-        VE_IGNORE_EINTR (logfd = open (log_path, O_CREAT|O_TRUNC|O_WRONLY|O_EXCL, 0644));
+        VE_IGNORE_EINTR (logfd = open (log_path, O_CREAT|O_APPEND|O_TRUNC|O_WRONLY|O_EXCL, 0644));
 
         g_free (log_path);
 
diff -up gdm-2.25.2/daemon/gdm-slave-proxy.c.append-logs gdm-2.25.2/daemon/gdm-slave-proxy.c
--- gdm-2.25.2/daemon/gdm-slave-proxy.c.append-logs	2009-01-22 10:44:33.915495761 -0500
+++ gdm-2.25.2/daemon/gdm-slave-proxy.c	2009-01-22 10:44:41.753495939 -0500
@@ -139,7 +139,7 @@ spawn_child_setup (SpawnChildData *data)
                 rotate_logs (data->log_file, MAX_LOGS);
 
                 VE_IGNORE_EINTR (g_unlink (data->log_file));
-                VE_IGNORE_EINTR (logfd = open (data->log_file, O_CREAT|O_TRUNC|O_WRONLY|O_EXCL, 0644));
+                VE_IGNORE_EINTR (logfd = open (data->log_file, O_CREAT|O_APPEND|O_TRUNC|O_WRONLY|O_EXCL, 0644));
 
                 if (logfd != -1) {
                         VE_IGNORE_EINTR (dup2 (logfd, 1));
diff -up gdm-2.25.2/daemon/gdm-welcome-session.c.append-logs gdm-2.25.2/daemon/gdm-welcome-session.c
--- gdm-2.25.2/daemon/gdm-welcome-session.c.append-logs	2009-01-22 10:44:05.300495464 -0500
+++ gdm-2.25.2/daemon/gdm-welcome-session.c	2009-01-22 10:44:14.903494653 -0500
@@ -482,7 +482,7 @@ spawn_child_setup (SpawnChildData *data)
                 rotate_logs (data->log_file, MAX_LOGS);
 
                 VE_IGNORE_EINTR (g_unlink (data->log_file));
-                VE_IGNORE_EINTR (logfd = open (data->log_file, O_CREAT|O_TRUNC|O_WRONLY|O_EXCL, 0644));
+                VE_IGNORE_EINTR (logfd = open (data->log_file, O_CREAT|O_APPEND|O_TRUNC|O_WRONLY|O_EXCL, 0644));
 
                 if (logfd != -1) {
                         VE_IGNORE_EINTR (dup2 (logfd, 1));
