--- trunk/gui/gdmcommon.c	2007/09/18 14:48:25	5279
+++ trunk/gui/gdmcommon.c	2007/09/18 18:39:42	5280
@@ -681,7 +681,9 @@
 	g_spawn_async (".",
 		       pre_fetch_prog_argv,
 		       NULL,
-		       (GSpawnFlags) (G_SPAWN_SEARCH_PATH),
+		       (GSpawnFlags) (G_SPAWN_SEARCH_PATH | 
+                              G_SPAWN_STDOUT_TO_DEV_NULL |
+                              G_SPAWN_STDERR_TO_DEV_NULL),
 		       NULL,
 		       NULL,
 		       &pid,
@@ -710,7 +712,9 @@
 	g_spawn_async (".",
 		       atspi_prog_argv,
 		       NULL,
-		       (GSpawnFlags) (G_SPAWN_SEARCH_PATH),
+		       (GSpawnFlags) (G_SPAWN_SEARCH_PATH |
+                              G_SPAWN_STDOUT_TO_DEV_NULL |
+                              G_SPAWN_STDERR_TO_DEV_NULL),
 		       NULL,
 		       NULL,
 		       &pid,
