diff -up gdm-2.20.0/gui/greeter/greeter_canvas_item.c.stupid-bullets gdm-2.20.0/gui/greeter/greeter_canvas_item.c
--- gdm-2.20.0/gui/greeter/greeter_canvas_item.c.stupid-bullets	2007-09-28 11:25:25.000000000 -0400
+++ gdm-2.20.0/gui/greeter/greeter_canvas_item.c	2007-09-28 11:26:08.000000000 -0400
@@ -469,7 +469,7 @@ greeter_item_create_canvas_item (Greeter
     else if (gdm_config_get_bool (GDM_KEY_ENTRY_CIRCLES))
       gtk_entry_set_invisible_char (GTK_ENTRY (entry), 0x25cf);
     else
-	    gtk_entry_set_invisible_char (GTK_ENTRY (entry), '*');
+	    gtk_entry_set_invisible_char (GTK_ENTRY (entry), 0x2022);
 
     gtk_widget_modify_font (entry, item->data.text.fonts[GREETER_ITEM_STATE_NORMAL]);
 
