diff -up gdm-2.23.92/gui/simple-greeter/gdm-user-manager.c.username-leak gdm-2.23.92/gui/simple-greeter/gdm-user-manager.c
--- gdm-2.23.92/gui/simple-greeter/gdm-user-manager.c.username-leak	2008-09-16 22:00:14.000000000 -0400
+++ gdm-2.23.92/gui/simple-greeter/gdm-user-manager.c	2008-09-16 22:00:36.000000000 -0400
@@ -1147,11 +1147,13 @@ process_ck_history_line (GdmUserManager 
         user = gdm_user_manager_get_user (manager, username);
         if (user == NULL) {
                 g_debug ("GdmUserManager: unable to lookup user '%s'", username);
+                g_free (username);
                 return;
         }
 
         g_object_set (user, "login-frequency", frequency, NULL);
         g_signal_emit (manager, signals [USER_LOGIN_FREQUENCY_CHANGED], 0, user);
+        g_free (username);
 }
 
 static gboolean
